drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1082 @@
+				fprintf(stdout, "crtc %u already connected to encoder %u\n",
+					crtc, conn->encoder_id);
+				drmModeFreeEncoder(enc);
+				dev->crtc_id = crtc;
+				return 0;
+			}
+		}
+
+		drmModeFreeEncoder(enc);
+	}
+
+	/* If the connector is not currently bound to an encoder or if the
+	 * encoder+crtc is already used by another connector (actually unlikely
+	 * but lets be safe), iterate all other available encoders to find a
+	 * matching CRTC. */
+	for (i = 0; i < conn->count_encoders; ++i) {
+		enc = drmModeGetEncoder(fd, conn->encoders[i]);
+		if (!enc) {
+			fprintf(stderr, "cannot retrieve encoder %u:%u (%d): %m\n",
+				i, conn->encoders[i], errno);
+			continue;
+		}
+
+		/* iterate all global CRTCs */
+		for (j = 0; j < res->count_crtcs; ++j) {
+			/* check whether this CRTC works with the encoder */
+			if (!(enc->possible_crtcs & (1 << j)))
+				continue;
+
+			/* check that no other device already uses this CRTC */
+			crtc = res->crtcs[j];
+			for (iter = modeset_list; iter; iter = iter->next) {
+				if (iter->crtc_id == crtc) {
+					crtc = -1;
+					break;
+				}
+			}
+
+			/* We have found a CRTC, so save it and return.
+			 * Note that we have to save its index as well.
+			 * The CRTC index (not its ID) will be used
+			 * when searching for a suitable plane.
+			 * TODO: the index is not really persistent?
+			 */
+			if (crtc >= 0) {
+				fprintf(stdout, "crtc %u found for encoder %u, will need full modeset\n",
+					crtc, conn->encoders[i]);;
+				drmModeFreeEncoder(enc);
+				dev->crtc_id = crtc;
+				dev->crtc_index = j;
+				return 0;
+			}
+		}
+
+		drmModeFreeEncoder(enc);
+	}
+
+	fprintf(stderr, "cannot find suitable crtc for connector %u\n",
+		conn->connector_id);
+	return -ENOENT;
+}
+
+static uint64_t get_property_value(int fd, drmModeObjectPropertiesPtr props, const char *name)
+{
+	drmModePropertyPtr prop;
+	uint64_t value;
+	int j;
+	
+	for (j = 0; j < props->count_props; j++) {
+		prop = drmModeGetProperty(fd, props->props[j]);
+		if (!strcmp(prop->name, name))
+			value = props->prop_values[j];
+		drmModeFreeProperty(prop);
+	}
+	return value;
+}
+
+/*
+ * TODO: need doc
+ */
+static int modeset_find_plane(int fd, struct modeset_dev *dev)
+{
+	drmModePlaneResPtr plane_res;
+	bool found_primary = false;
+	int i, ret = -EINVAL;
+
+	plane_res = drmModeGetPlaneResources(fd);
+	if (!plane_res) {
+		fprintf(stderr, "drmModeGetPlaneResources failed: %s\n", strerror(errno));
+		return -ENOENT;
+	}
+
+	for (i = 0; (i < plane_res->count_planes) && !found_primary; i++) {
+		int plane_id = plane_res->planes[i];
+
+		drmModePlanePtr plane = drmModeGetPlane(fd, plane_id);
+		if (!plane) {
+			fprintf(stderr, "drmModeGetPlane(%u) failed: %s\n", plane_id, strerror(errno));
+			continue;
+		}
+
+		if (plane->possible_crtcs & (1 << dev->crtc_index)) {
+			drmModeObjectPropertiesPtr props =
+				drmModeObjectGetProperties(fd, plane_id, DRM_MODE_OBJECT_PLANE);
+
+			/* primary or not, this plane is good enough to use. */
```

Comment:
I doubt that all hardware would be fine with running a CRTC without a primary plane specifically.